### PR TITLE
Piro: Use reduced Obj for sensitivity computations and reduced Hessian-vector products.

### PR DESCRIPTION
**Pull request by kliegeois**
@trilinos/rol 
@mperego 

## Motivation

This PR is intended to add capabilities in Piro to use reduced objective function to compute both sensitivities and reduced Hessian-vector products.

This PR impacts both Piro and ROL:

ROL: 
 - The Thyra ROL adapters have been updated to be defined inside the ROL namespace,
 - The member functions and data of the Reduced Objective class have been declared as `protected` instead of `private`; this allows to define a child class in Piro which can access those functions and data.

Piro:
 - Support for `OldReducedSpace` has been dropped;
 - Support for Epetra with NOX has been dropped;
 - A reduced objective class has been defined such that precomputed solution can be set;
 - If Piro is compiled with ROL support, the sensitivities are computed using the above-mentioned child class;
  - If Piro is compiled with ROL support, the reduced Hessian-vector products can be computed with the above-mentioned child class;
  - A test, `Piro_ThyraSolverTpetra`, has been introduced to test the reduced Hessian-vector product computations.

Trilinos tests passed with this branch on Blake.

### CODE REVIEW

File: packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp [hunk 1]
```diff
@@ -54,15 +54,15 @@
...
+namespace ROL {
 
 //! \brief ROL interface wrapper for Sacado SimOpt Constraint
 template<class Real>
 class ThyraProductME_Constraint_SimOpt : public Constraint_SimOpt<Real> {
 
 public:
 
-  ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
+  ThyraProductME_Constraint_SimOpt(const Thyra::ModelEvaluator<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
```
Comment (mperego) [inline packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65]: why did you have to change this?

File: packages/piro/test/_input_Analysis_ROL_AdjointSensitivities_OldReducedSpace_Tpetra.xml [hunk 2]
```diff
@@ -151,8 +151,6 @@
         <Parameter name="Error File" type="string" value="dak.err" />
       </ParameterList>
       <ParameterList name="ROL">
```
Comment (mperego) [inline packages/piro/test/_input_Analysis_ROL_AdjointSensitivities_OldReducedSpace_Tpetra.xml:153]: We should probably remove or rename (if it does something different than the other tests) this file

File: packages/piro/test/_input_Analysis_ROL_OldReducedSpace_Tpetra.xml [hunk 3]
```diff
@@ -155,7 +155,6 @@
         <Parameter name="Error File" type="string" value="dak.err" />
       </ParameterList>
       <ParameterList name="ROL">
-        <Parameter name="Use Old Reduced Space Interface" type="bool" value="true" />
```
Comment (mperego) [inline packages/piro/test/_input_Analysis_ROL_OldReducedSpace_Tpetra.xml:158]: same here

File: packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective.hpp [hunk 4]
```diff
@@ -62,7 +62,7 @@ template <class Real>
 class ThyraProductME_Objective : public Objective<Real> {
 public:
 
-  ThyraProductME_Objective(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
+  ThyraProductME_Objective(Thyra::ModelEvaluator<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
```
Comment (mperego) [inline packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective.hpp:65]: same question here.

File: packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp [hunk 5]
```diff
@@ -50,15 +50,15 @@
...
+namespace ROL {
 
 template <class Real>
 class ThyraProductME_Objective_SimOpt : public Objective_SimOpt<Real> {
 
 public:
 
 
-  ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
+  ThyraProductME_Objective_SimOpt(const Thyra::ModelEvaluator<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
```
Comment (mperego) [inline packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61]: and here

**Review by mperego — COMMENTED** [review]
It looks good. Thanks for all you work. 
I only had a few comments/questions

File: packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp (hunk 1 above)
Comment (kliegeois) [reply packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65]: It was required to compile the `evalConvergedModelResponsesAndSensitivities` function.

Not doing those modifications leads to the following errors:

```
[ 85%] Built target ROL_test_compatibility_stacktrace
[ 85%] Building CXX object packages/piro/src/CMakeFiles/piro.dir/Piro_NOXSolver.cpp.o
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp: In instantiation of ‘void Piro::SteadyStateSolver<Scalar>::evalConvergedModelResponsesAndSensitivities(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:204:3:   required from ‘void Piro::NOXSolver<Scalar>::evalModelImpl(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:50:16:   required from here
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:431:52: error: no matching function for call to ‘ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
       ROL::ThyraProductME_Objective_SimOpt<Scalar> obj(*model_, i, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                    ^~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:79:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note: candidate: ROL::ThyraProductME_Objective_SimOpt<Real>::ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const ROL::ThyraProductME_Objective_SimOpt<double>&)
 class ThyraProductME_Objective_SimOpt : public Objective_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(ROL::ThyraProductME_Objective_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:432:53: error: no matching function for call to ‘ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
       ROL::ThyraProductME_Constraint_SimOpt<Scalar> constr(*model_, i, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                     ^~~~~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:80:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<Real>::ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const ROL::ThyraProductME_Constraint_SimOpt<double>&)
 class ThyraProductME_Constraint_SimOpt : public Constraint_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(ROL::ThyraProductME_Constraint_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp: In instantiation of ‘void Piro::SteadyStateSolver<Scalar>::evalReducedHessian(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:205:3:   required from ‘void Piro::NOXSolver<Scalar>::evalModelImpl(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:50:16:   required from here
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:1218:50: error: no matching function for call to ‘ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
     ROL::ThyraProductME_Objective_SimOpt<Scalar> obj(*model_, g_index, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                  ^~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:79:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note: candidate: ROL::ThyraProductME_Objective_SimOpt<Real>::ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const ROL::ThyraProductME_Objective_SimOpt<double>&)
 class ThyraProductME_Objective_SimOpt : public Objective_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(ROL::ThyraProductME_Objective_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:1219:51: error: no matching function for call to ‘ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
     ROL::ThyraProductME_Constraint_SimOpt<Scalar> constr(*model_, g_index, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                   ^~~~~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:80:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<Real>::ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const ROL::ThyraProductME_Constraint_SimOpt<double>&)
 class ThyraProductME_Constraint_SimOpt : public Constraint_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(ROL::ThyraProductME_Constraint_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
packages/piro/src/CMakeFiles/piro.dir/build.make:206: recipe for target 'packages/piro/src/CMakeFiles/piro.dir/Piro_NOXSolver.cpp.o' failed
make[2]: *** [packages/piro/src/CMakeFiles/piro.dir/Piro_NOXSolver.cpp.o] Error 1
CMakeFiles/Makefile2:76691: recipe for target 'packages/piro/src/CMakeFiles/piro.dir/all' failed
make[1]: *** [packages/piro/src/CMakeFiles/piro.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2

```

I think that those modifications are fine; they increase the range of applicability of those 3 objects.

File: packages/piro/test/_input_Analysis_ROL_OldReducedSpace_Tpetra.xml (hunk 3 above)
Comment (kliegeois) [reply packages/piro/test/_input_Analysis_ROL_OldReducedSpace_Tpetra.xml:158]: Same answer as above.

File: packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective.hpp (hunk 4 above)
Comment (kliegeois) [reply packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective.hpp:65]: Same answer as above.

File: packages/piro/test/_input_Analysis_ROL_AdjointSensitivities_OldReducedSpace_Tpetra.xml (hunk 2 above)
Comment (kliegeois) [reply packages/piro/test/_input_Analysis_ROL_AdjointSensitivities_OldReducedSpace_Tpetra.xml:153]: I think that this test is redundant with one other test now.
I have to check that.

**Comment by kliegeois** [comment]
Thanks @mperego for your review!

**Comment by kliegeois** [comment]
@mperego , I just deleted two unused tests and renamed the ones related to the `OldReducedSpace`.
I verify that the remaining tests are not duplicated.

**Review by mperego — APPROVED** [review]
Thanks for renaming/removing the tests.
IT looks good to me!

**Pull request merged**